pyClient/zethTestScenario.py

```diff
@@ -85,8 +86,8 @@ def bob_to_charlie(test_grpc_endpoint, mixer_instance, mk_root, mk_path1, input_
 
     output_note1_str = json.dumps(zethGRPC.parseZethNote(output_note1))
     output_note2_str = json.dumps(zethGRPC.parseZethNote(output_note2))
-    ciphertext1 = zethUtils.encrypt(output_note1_str, keystore["Bob"]["AddrPk"]["ek"]) # Bob is the recipient
-    ciphertext2 = zethUtils.encrypt(output_note2_str, keystore["Charlie"]["AddrPk"]["ek"]) # Charlie is the recipient
+    ciphertext1 = zethUtils.encrypt(output_note1_str, keystore["Bob"]["AddrPk"]["encPK"], keystore["Bob"]["AddrSk"]["encSK"]) # Bob is the recipient
+    ciphertext2 = zethUtils.encrypt(output_note2_str, keystore["Charlie"]["AddrPk"]["encPK"], keystore["Bob"]["AddrSk"]["encSK"]) # Charlie is the recipient
 
     return zethContracts.mix(
```

Comment:
Looks like the sender's (lets call the sender Alice) PK is not provided as an argument to the `mix` call. This means that, the only chance Bob (the recipient here) has to recover the note plaintext is if he knows he is the recipient of a transaction from Alice.
This posses several issues because:
- Bob might be the recipient of several senders
- Bob does not know if which order the tx will be mined (hence cannot prepare the decryption keys in the correct order to scan the chain)
- If Bob's hardware burns, and Bob has done a great job at backing up his keys, then Bob will still not be able to re-scan the chain to recover the received notes..

As a result, Bob needs to create the entire set of keys derived from his secret and all potential senders keys and scan the chain with this set of keys.. That's quite unusable.

To address that, we should add the PK of the sender to the tx object so that Bob can do a DH locally to recover the decryption key that will enable him to recover the note's plaintexts.

Let's think about key rotation and forward secrecy too.

Reply:
The option to add the sender PK key looks ok to me since it would avoid the need to brute-force all the possible senders. So I guess I could add it as part of the log. As we discussed offline, ZCash adopt a similar solution (4.16.1 https://github.com/zcash/zips/blob/master/protocol/protocol.pdf).

About the forward/backward secrecy the problem is that the shared secret (
symmetric one) is always the same. If an attacker is able to recover it, he is able to decrypt all previous/future ciphertexts. An idea, so, would be to generate a different shared key for each encryption, this can be achieved by using ephemeral dh keypairs for the sender. In Zcash this approach seems to be used mainly for sender anonymity. In our case, it will be used mainly for forward secrecy in case a single shared key is leaked. It worth mentioning that if the private key of the receiver is leaked we do not have neither backward nor forward secrecy since the attacker is able to recover all the different shared keys. Key rotation (changing `enc_sk`,`enc_pk` each time T) could partially solve this problem allowing "time windows" of backward/forward secrecy: what I mean is if `(sk_T)` is leaked we have that for t<=`T-1` and t>=`T+1` ciphertexts are still secure since derived by different private key.

Reply:
> Looks like the sender's (lets call the sender Alice) PK is not provided as an argument 
> The option to add the sender PK

I just want to make sure that you are talking about not the long standing sender's PK but the ephemeral one-time `epk` as specified in §4.16 of the Zcash sepcification doc?

Reply:
Yes, having ephemeral key pairs is what we want as it improves forward/backward secrecy (note that anonymity is not provided by Zeth since we need to pay for the gas to execute the smart contract, so this is not for this direct reason that we are interested in having ephemeral keys). We do not have the choice anyway to add the PK - `epk` - used by the sender for the transaction as part of the tx object, since otherwise it becomes unpractical to recover the plaintext of the notes, and to transact through the Mixer contract.

Reply:
Merge request now supports ephemeral public key for the sender.